File: scripts/private_root.py
import os
from pathlib import Path

PRIVATE_ROOT_ENV = "ASHEBROOK_PRIVATE_ROOT"
REPOSITORY_ROOT = Path(__file__).resolve().parent.parent


def resolve_private_root(repository_root=REPOSITORY_ROOT, configured_root=None):
    configured = (
        os.environ.get(PRIVATE_ROOT_ENV, "")
        if configured_root is None
        else configured_root
    )
    selected = configured.strip() or "private"
    root = Path(selected)
    if not root.is_absolute():
        root = Path(repository_root) / root
    return root.resolve()


def private_path(*segments):
    return resolve_private_root().joinpath(*segments)

File: scripts/private-root.ts
import path from 'node:path';
import { fileURLToPath } from 'node:url';

export const PRIVATE_ROOT_ENV = 'ASHEBROOK_PRIVATE_ROOT';

export function repositoryRootFromModuleUrl(
  moduleUrl: string,
  fallbackRoot = process.cwd(),
): string {
  const url = new URL(moduleUrl);
  if (url.protocol !== 'file:') return path.resolve(fallbackRoot);
  return path.resolve(path.dirname(fileURLToPath(url)), '..');
}

export const REPOSITORY_ROOT = repositoryRootFromModuleUrl(import.meta.url);

export interface PrivateRootOptions {
  repositoryRoot?: string;
  configuredRoot?: string;
}

export function resolvePrivateRoot(options: PrivateRootOptions = {}): string {
  const repositoryRoot = options.repositoryRoot ?? REPOSITORY_ROOT;
  const configuredRoot =
    options.configuredRoot ?? process.env[PRIVATE_ROOT_ENV] ?? '';
  const selectedRoot = configuredRoot.trim() || 'private';
  return path.resolve(repositoryRoot, selectedRoot);
}

export function resolvePrivatePath(
  segments: readonly string[],
  options: PrivateRootOptions = {},
): string {
  return path.resolve(resolvePrivateRoot(options), ...segments);
}
